File: java_OOP/generic_purpose/src/Deque.java
interface Deque<E> extends Stack<E>, Queue<E> {
	void addFirst(E element); // リストの先頭に要素を追加します。

	void addLast(E element); // リストの末尾に要素を追加します。

	E removeFirst(); // リストの最初の要素を削除し、削除した要素を返します。

	E removeLast(); // リストの最後の要素を削除し、削除した要素を返します。

	int size(); // リストの要素数を返します。
}

File: php/src/oop/src/Models/Deque/Deque.php
<?php

declare(strict_types=1);

namespace App\Models\Deque;

/**
 * Deque インターフェース (両端キュー - Double-ended queue)
 *
 * Deque インターフェースは、Stack と Queue の両方から拡張されます。
 * Deque は、Stack と Queue のすべての振る舞いを持ち、
 * さらにリストの前に要素を追加する能力（addFirst）を持つインターフェースです。
 *
 * PHPでは、インターフェースの複数継承が可能です。
 */
interface Deque extends Stack, Queue
{
    /**
     * 要素をリストの最初に追加します
     *
     * @param int $value 追加する値
     * @return void
     */
    public function addFirst(int $value): void;
}
